Docker & CI build: improve speed by leveraging caching

derhuerst opened the pull request:
Still missing:

- [ ] ~~Mount `.yarn/cache` into `node-modules` the Docker build.~~
- [x] Try caching the `node-modules` stage, and check if it is actually faster than building from scratch. – much faster, I didn't expect that
- [x] Check if the CI runs as intended.
- [ ] Investigate why `yarn build` is a lot slower than before. 🤔 ([ ~5m before](https://github.com/HSLdevcom/digitransit-ui/runs/6112614393?check_suite_focus=true#step:8:1), 10-20m after) – due to the upgraded `favicons-webpack-plugin`
- [x] My addition of `--immutable` to `yarn install` causes the build to fail, as currently both on `v3` and `improve-ci-speed`, the checked-in `yarn.lock` does not reflect the workspace packages' versions. cc @vesameskanen

## Proposed Changes

This PR changes the `Dockerfile` & GitHub Actions dev pipeline to be faster.

While an un-cached (different `yarn.lock` and/or different `digitransit-*/**` and/or different commit hash) build is *slightly* slower, a cached one is is much faster.

## Review

For an unknown reason, `favicons-webpack-plugin@0.0.9` crashed during the Docker build, and `favicons-webpack-plugin@^4` froze, so I used `favicons-webpack-plugin@^3`.

When looking at the changes, I recommend to review each commit individually, as I tried to make them as atomic and sensible as possible. Some of them also contain explanations.

Unfortunately, because [the GitHub Actions system seems to lack support](https://github.com/orgs/community/discussions/25545) for [YAML anchors](https://yaml.org/spec/1.2.2/#example-node-for-sammy-sosa-appears-twice-in-this-document), I added a lot of duplicated code in `.github/workflows/dev-pipeline.yml`. The benefit of the repeated "use cache if exists, otherwise rebuild" patterns is that each step is effectively idempotent – it works regardless of wether the cache is available and if a previous step populated it –, with the downside being very repetitive code. Once YAML anchors are supported, we could DRY it a lot.

derhuerst committed:
Dockerfile: minor tweaks

- use 1.4 syntax to support `COPY --link` commands
- set NODE_ENV=production
- remove $WORK build arg

derhuerst committed:
Dockerfile: use multi-stage build for yarn install

derhuerst committed:
Dockerfile: yarn install: run with --immutable

derhuerst committed:
Dockerfile: yarn install: tweak CLI output

derhuerst committed:
Dockerfile: node:12 -> node:14-alpine, CI dev pipeline: use Node v14

derhuerst committed:
Dockerfile: yarn build: tweak removal of caches & artifacts, tweak caching

- If e.g. `test` has been added in a command/step, removing it in a
subsequent command/step does not have a positive effect on the image size.
Rather, we add it .dockerignore to stop adding it in the first place.
- node_modules/.cache & /tmp/Relay* are populated by Webpack plugins.

derhuerst committed:
fix Docker build: favicons-webpack-plugin@^3 & update webpack@^4

derhuerst committed:
make source map generation configurable

derhuerst committed:
CI dev pipeline: cache .yarn/cache in addition to node_modules

Even if a commit has changed `yarn.lock`, we can use an old `.yarn/cache` directory, so that we don't have to fetch 99% of all dependencies again.

derhuerst commented on .github/workflows/dev-pipeline.yml:
```diff
@@ -3,6 +3,8 @@ on:
   push:
     branches:
       - v2
+      # todo: remove
```
This needs to be removed before merging!

derhuerst commented on .github/workflows/scripts/build_and_push_dev.sh:
```diff
@@ -15,10 +15,13 @@ echo "Building digitransit-ui"
-docker tag $DOCKER_IMAGE_TAG_LONG $DOCKER_IMAGE_TAG
-docker push $DOCKER_IMAGE_TAG
+# ---
+# ONLY re-add these lines after removing the `improve-ci-speed` branch opt-in the GH Actions workflow!!
```
This needs to be removed before merging!

derhuerst committed:
.dockerignore: ignore .github, keep README & co, remove obsolete paths

derhuerst committed:
Webpack stats.json: only include .entrypoints

app/server.js reads .entrypoints.main.assets in order to generate an appropriate
manifest and `preload` tags, but it doesn't need the rest. This is why we turn
all stats off via `all: false` and then selectively turn `entrypoints` on.

Before, stats.json was 82mb, now it is ~9kb.

derhuerst committed:
Dockerfile: add labels following OCI image spec

derhuerst commented on .github/workflows/dev-pipeline.yml:
```diff
@@ -225,14 +416,17 @@ jobs:
   publish-npm:
-    if: github.ref == 'refs/heads/v2'
+    # todo: re-add this before merging!
+    # if: github.ref == 'refs/heads/v2'
```
This needs to be removed before merging!

derhuerst commented on .github/workflows/dev-pipeline.yml:
```diff
@@ -262,24 +483,26 @@ jobs:
-          PUBLISHED_PACKAGES: ${{ steps.log-parse.outputs.published-packages }}
-        shell: bash
+      # todo: re-add this before merging!
+      # - name: Publish shared components to npm
```
This needs to be removed before merging!

derhuerst commented on .github/workflows/dev-pipeline.yml:
```diff
@@ -262,24 +483,26 @@ jobs:
   docker-push:
-    if: github.ref == 'refs/heads/v2'
+    # todo: re-add this before merging!
+    # if: github.ref == 'refs/heads/v2'
```
This needs to be removed before merging!

derhuerst committed:
CI dev pipeline: re-install node_modules only if not cached

derhuerst committed:
CI dev pipeline: rebuild shared components only if not cached

derhuerst committed:
CI dev pipeline: rebuild Relay queries only if not cached

derhuerst committed:
CI dev pipeline: don't let Playwright download unused browsers

derhuerst committed:
CI dev pipeline: minor tweaks

derhuerst committed:
CI dev/hotfix pipelines: build & publish Docker img via buildx & GH action

derhuerst committed:
CI dev/hotfix pipelines: cache Docker layers via GitHub cache

derhuerst committed:
CI: merge v3-dev into dev pipeline

derhuerst committed:
Dockerfile: yarn install: skip Playwright browser download

derhuerst committed:
debug CI [drop]

derhuerst committed:
CI: use env var & $GITHUB_OUTPUT instead of set-output

set-output has been deprecated: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

derhuerst committed:
jyvaskyla config.logo: use relative path

This probably improves favicons-webpack-plugin's performance, as node_modules won't be resolved for resolving the path.

derhuerst committed:
Dockerfile: add health check

derhuerst committed:
Dockerfile: allow passing in $CONFIG as build arg

derhuerst committed:
Yarn: run Node v14 patch after each install

vesameskanen commented:
It seems that the new favicon setup is broken. I built the UI server locally (yarn build), and run it with a config, for example:

CONFIG=tampere NODE_OPTS='--title=digitransit-ui' yarn start

Chrome debug tools indicate that no favicons are found (application/manifest). Broken paths  in the manifest are of type:
 
http://localhost:8080/icons/android-chrome-36x36.png

Correct icon path from v3 branch looks like:

http://localhost:8080/assets/icons-tampere-70020a06183cc7648f25613739dcb137/android-chrome-36x36.png

vesameskanen commented:
A further note: all required favicons are generated, only the paths are wrong.

hbruch commented:
Upgrading `webpack-assets-manifest` to `^4.0.0` (with `html-webpack-plugin` resolved to `4.5.2`) recreated the correct icon paths in the `manifest.json` (see https://github.com/embarkokc/digitransit-ui/pull/26)

derhuerst commented:
> It seems that the new favicon setup is broken. […] Chrome debug tools indicate that no favicons are found (application/manifest). […]

It seems like the HTTP request for the favicon is redirected wrongly:

```shell
curl 'http://localhost:8080/icons/android-chrome-36x36.png' -L -I
# HTTP/1.1 302 Found
# X-Powered-By: Express
# Set-Cookie: lang=fi; Path=/
# Location: /reitti/icons/android-chrome-36x36.png
# Vary: Accept
# Content-Type: text/plain; charset=utf-8
# Content-Length: 60
# Date: Wed, 29 Mar 2023 12:45:18 GMT
# Connection: keep-alive
# Keep-Alive: timeout=5

# HTTP/1.1 200 OK
# X-Powered-By: Express
# Set-Cookie: lang=fi; Path=/
# content-type: text/html; charset=utf-8
# Date: Wed, 29 Mar 2023 12:45:18 GMT
# Connection: keep-alive
# Keep-Alive: timeout=5
```

vesameskanen commented:
Right. Correct path would be `assets/icons-<theme>-<hash>/<imagename>`, not` icons/<imagename>`. 

Yarn build creates all required icons into the correct path, but somehow the build ends up configuring  wrong path.

derhuerst committed:
CI dev/hotfix pipeline: re-enable linting of SVG icons

This reverts 32bbfa8, which reverts 3f6e105, which apparently failed because of an unrelated problem[1].

[1] https://github.com/HSLdevcom/digitransit-ui/actions/runs/4762670497/jobs/8469259999

derhuerst committed:
don't build startup images for iOS/iPadOS

This speeds up the webpack build within the Docker image build significantly.

derhuerst commented:
In 6db1b69f1, I had changed the `yarn install` to use [a Docker cache mount (`RUN --mount=type=cache`)](https://docs.docker.com/engine/reference/builder/#run---mounttypecache), but it doesn't seem to work with GitHub Actions, failling with `rm -rf node_modules/.cache: resource busy`. I have removed it from this PR.
